### PR TITLE
Import a single Gist by ID

### PR DESCRIPTION
**Pull request by jazcarate**
There are 2 "discussion points":

## Limitations
1. Adding a gist with files larger than 10mb
2. Adding a gist with more than 300 files

The `1` is easy enough, as you can fetch the `raw_url` direct.
The `2` is a bit tricky, as one should clone the git repo. I was not about to make the server clone a repo to get the 301th file 😅 so I opted to simply `throw`. We can improve on it later 🤷 

### Gists to test these limits:

1. Big file: https://gist.github.com/jazcarate/50bb04ad4904dc42dce512c10e9e0557
2. Many files: https://gist.github.com/jazcarate/bf8b6ac8c82f09946e350814aa3c84a1

## Creation date
I opted to import the creation date of a gist, but not the `updated_at`. It is a bit weird that you can import an old gist, and not see it in the `/mine`, but I felt it was more convenient to keep the `created_at`. We can make the creation date be the imported date. I don't have a good reason for either one 😸 

# Pending
A UI. I wanted to push this first part, discuss the endpoint, then make a bulk import (maybe a queue? so we don't overload the server. Or even push the `fetching` to the frontend 🤔)

### CODE REVIEW

File: server/src/lib/gist/fetch.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import { Response } from "node-fetch"
+import { Gist, GistFile } from "./types"
+
+async function fetchHelper(response: Response): Promise<Response> {
+	if (!response.ok) {
+		const isJson = response.headers
+			.get("content-type")
+			?.includes("application/json")
+		const err = await (isJson ? response.json() : response.text())
+		console.warn("Couldn't fetch because", err)
```
Comment (MaxLeiter) [inline server/src/lib/gist/fetch.ts:11]: Do we need to print this on the server? I think its enough to send it (or at least the message) back to the client. Maybe wrap it in a `NODE_ENV === 'development'`?

File: server/src/lib/gist/fetch.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		content: file.truncated
+			? () =>
+					fetch(file.raw_url)
+						.then(fetchHelper)
+						.then((res) => res.text())
+			: () => Promise.resolve(file.content)
+	}
+}
+
+export function toGist(response: GistResponse): Gist {
```
Comment (MaxLeiter) [inline server/src/lib/gist/fetch.ts:47]: nitpick: can we call it `responseToGist` or something more descriptive?

File: server/src/routes/posts.ts [hunk 3]
```diff
@@ -374,3 +375,44 @@ posts.delete("/:id", jwt, async (req: UserJwtRequest, res, next) => {
...
+	}),
+	async (req: UserJwtRequest, res, next) => {
+		try {
+			const { id } = req.params
+			const { visibility, password, expiresAt } = req.body
+			const gist = await getGist(id)
+
+			let hashedPassword: string = ""
+			if (visibility === "protected") {
+				hashedPassword = crypto
```
Comment (MaxLeiter) [inline server/src/routes/posts.ts:399]: we have this `crypto.createHash...` line a few places. Would you mind refactoring it to a `lib/generate-password` or something? DRY and all that

File: server/src/lib/gist/__tests__/index.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
+import { Post } from "@lib/models/Post"
```
Comment (MaxLeiter) [inline server/src/lib/gist/__tests__/index.ts:1]: oh cool, I wonder who added the testing framework.  🙏

**Review by MaxLeiter — DISMISSED** [review]

**Comment by MaxLeiter** [comment]
Thanks for this! 

What's the problem with /mine and created_at? I think it makes sense to set created_at to the original Gist date and updated_at to the time of the import. Then we can change `/mine` to sort by updatedAt || createdAt and `/mine` should be accurate, right?

File: server/src/lib/gist/__tests__/index.ts (hunk 4 above)
Comment (jazcarate) [reply server/src/lib/gist/__tests__/index.ts:1]: 😊

File: server/src/routes/posts.ts (hunk 3 above)
Comment (jazcarate) [reply server/src/routes/posts.ts:399]: Actually, I wanted to restructure the whole routes into _thinner_ layers, and not cram domain logic, with DB with controller 😅. That's why I opted for the simplest (copy & paste) and discuss elsewhere how we could split the responsibilities. I created this proposal: https://github.com/MaxLeiter/Drift/issues/79

**Comment by jazcarate** [comment]
> What's the problem with /mine and created_at?

Not a problem, so much as weird UX. Say you have an old Gist, and you created 10 Drift Posts.
1. You go to `/mine` you see your 10 posts
2. You import the old gist
3. You go back to `/mine` and **nothing changed** (because the Gist is older than the 10 posts)

Sorting by `createdAt` does _fix_ this UX, but then your `mine` would be a bit disorganized. Every update to a drift would reorder the `/mine`.

Given that importing a gist should be much less common than updating an existing post; I left the `createdAt` as the Gist's creation date. But I wanted to point out the scenario ☝️. Maybe there is a better solution, like having an aditional `DriftCreationAt` that is used to sort `/mine`, but it is always the `new Date()` regardless of a creation or an import🤷.
TBH, I'd leave this as is, and see if anyone complains in the future when importing Gists 😸

**Comment by MaxLeiter** [comment]
> Sorting by createdAt does fix this UX, but then your mine would be a bit disorganized. Every update to a drift would reorder the /mine. 

Do you mean `sorting by updatedAt`? I think it's fine if modifying the visibility changes it in /mine.

Alternatively, we add a `source` string (open to other names) to the Post model that can for now be `"drift" | "gist"`. In the future this could be nice if someone uses a tool or API key to create a post.

**Comment by jazcarate** [comment]
> Do you mean `sorting by updatedAt`?
Yeah 😅


I wouldn't add a `source` _yet_. And I'll work on a restructure of the server/ after this get merged :)

**Review by MaxLeiter — APPROVED** [review]

**Pull request merged**